### PR TITLE
support xparl in notebook

### PR DESCRIPTION
**Pull request by zenghsh3**

### CODE REVIEW

File: parl/remote/utils.py [hunk 1]
```diff
@@ -52,49 +54,115 @@ def data_process():
     return to_write_lines
 
 
-def load_remote_class(file_name, class_name, end_of_file, in_sys_path):
+def is_class_defined_in_file(cls):
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/utils.py]: is_class_defined_in_file -> is_implemented_in_notebook.

and update the comment: 
check if the remote class is implemented in notebook environment.

File: parl/remote/utils.py [hunk 2]
```diff
@@ -52,49 +54,115 @@ def data_process():
...
+    """Check whether the class is defined in a python file
+
+    Args:
+        cls: class
     """
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/utils.py]: Please assert that the passed argument is a `class`.
https://stackoverflow.com/questions/395735/how-to-check-whether-a-variable-is-a-class-or-not

File: parl/remote/utils.py [hunk 3]
```diff
@@ -52,49 +54,115 @@ def data_process():
...
+        cls_module = sys.modules.get(cls.__module__)
+        if getattr(cls_module, '__file__', None):
+            return True
+    return False
 
-  Args:
-    file_name: specify the file to load the class
-    class_name: specify the class to be loaded
-    end_of_file: line ID to indicate the last line that defines the class.
-    in_sys_path: whether the path of the remote class is in the environment path (sys.path).
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/utils.py]: There are too many functions in `utils.py`. 
I suggest opening a new file named `remote_class_distribution` for these implementation.

File: parl/remote/utils.py [hunk 4]
```diff
@@ -52,49 +54,115 @@ def data_process():
...
+            mod = __import__(new_file_name)
+        else:
+            module_name = module_name.lstrip('.' + os.sep).replace(os.sep, '.')
+            mod = __import__(module_name, globals(), locals(), [class_name], 0)
+
+        cls = getattr(mod, class_name)
+
+        os.remove(tmp_file_name)
+    else:
+        cls = dumped_class_info
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/utils.py]: I'm afraid that the remote class distributed by cloudpickle does not support the newest feature.
The users can access the attributes of remote instances, but it is not clear if another implementation supports this new feature.

**Comment by TomorrowIsAnOtherDay** [comment]
Let's discuss how to add unit tests for the PR. Some code has not yet been covered.

File: parl/remote/utils.py (hunk 1 above)
Comment (zenghsh3) [reply parl/remote/utils.py]: check if the remote class is implemented in the environments like notebook(e.g., ipython, notebook).

File: parl/remote/utils.py (hunk 3 above)
Comment (zenghsh3) [reply parl/remote/utils.py]: remote_class_serialization

File: parl/remote/utils.py (hunk 1 above)
Comment (zenghsh3) [reply parl/remote/utils.py]: fixed

File: parl/remote/utils.py (hunk 2 above)
Comment (zenghsh3) [reply parl/remote/utils.py]: fixed

File: parl/remote/utils.py (hunk 3 above)
Comment (zenghsh3) [reply parl/remote/utils.py]: fixedd

File: parl/remote/utils.py (hunk 4 above)
Comment (zenghsh3) [reply parl/remote/utils.py]: added unittest

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM.

**Pull request merged**